### ISSUE TITLE
Span reporter can block the whole app in if proxy is not available

### ISSUE THREAD

**Issue by alexey-stankevich** [wavefrontHQ/wavefront-opentracing-sdk-python#59]
The WavefrontSpanReporter uses a queue with the size limited to 50000 by default to store spans not yet sent to Wavefront. However, when it puts a new span to this queue, it does so in the blocking mode (which is the default for the Python queue implementation). In case if the queue is not consumed (proxy is not available for example), the span.close() method blocks the whole application when the limit is reached. Here's the code to reproduce it:

```python
from wavefront_opentracing_sdk import WavefrontTracer
from wavefront_opentracing_sdk.reporting import WavefrontSpanReporter
from wavefront_sdk import WavefrontProxyClient
from wavefront_sdk.common import ApplicationTags

sender = WavefrontProxyClient(
    host='broken',
    tracing_port=30000,
    distribution_port=2878,
    metrics_port=2878
)

def getTracer():
    applicationTags = ApplicationTags(application='TestAPP', service='test')
    spanReporter = WavefrontSpanReporter(client=sender, max_queue_size=1, log_percent=1)
    tracer = WavefrontTracer(
        reporter=spanReporter,
        application_tags=applicationTags
    )
    return tracer


if __name__ == '__main__':
    tracer = getTracer()
    for i in range(3):
        with tracer.start_active_span('test operation') as scope:
            pass
```

This piece of code will never finish.

Note that this behavior is different from the one described in the docstring:

>  :param max_queue_size: Max queue size of in-memory buffer. Incoming
>         spans are dropped if buffer is full.

**Comment by oppegard** [comment]
Thanks for reporting this. We're having someone on the team investigate and they'll follow up with you in the next few days.

**Comment by alexey-stankevich** [comment]
Hello,

I've just validated my snippet and it still hangs. I'm using 1.2.11 with Python 2.7 (unfortunately we're still have to use it).

**Comment by keep94** [comment]
If you have to use python 2.7, I am guessing you are unable to upgrade to 1.8.2 of wavefront-sdk-python?

**Comment by alexey-stankevich** [comment]
Yes, we can't upgrade, that's why for now we use custom build of opentracing-sdk. But I've just checked with the latest versions and Python 3.9, and the issue is still there:
```
➜  ~ pip3 list | grep wavefront
wavefront-opentracing-sdk-python  2.2.2
wavefront-pyformance              1.1.0
wavefront-sdk-python              1.8.2
```
```
DeprecationWarning: Call to deprecated class WavefrontProxyClient.
  sender = WavefrontProxyClient(
WARNING:root:Can not report ~component.heartbeat to wavefront
Exception in thread wavefront_span_reporter:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/common/proxy_connection_handler.py", line 60, in send_data
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/common/proxy_connection_handler.py", line 45, in connect
    self._reconnecting_socket.connect((self._address, self._port))
socket.gaierror: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.9/site-packages/wavefront_opentracing_sdk/reporting/wavefront.py", line 60, in run
    self.send(wavefront_span)
  File "/usr/local/lib/python3.9/site-packages/wavefront_opentracing_sdk/reporting/wavefront.py", line 70, in send
    self.sender.send_span(
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/proxy.py", line 348, in send_span
    raise error
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/proxy.py", line 342, in send_span
    self._tracing_proxy_connection_handler.send_data(line_data)
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/common/proxy_connection_handler.py", line 67, in send_data
    self.send_data(line_data, reconnect=False)
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/common/proxy_connection_handler.py", line 71, in send_data
    raise error
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/common/proxy_connection_handler.py", line 60, in send_data
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/wavefront_sdk/common/proxy_connection_handler.py", line 45, in connect
    self._reconnecting_socket.connect((self._address, self._port))
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```
There's no record 
```
ERROR:root:Buffer full, dropping span: test operation, 956c5ca0-bc92-11ec-92e1-acde48001122
WARNING:root:Total spans dropped: 1
```
which is there when the `put` method is called with `block=False`, so I'd guess that missing of this argument is just a mistake (wrong default value assumed?)

**Comment by keep94** [comment]
Thank you for the information. So it sounds like the code snippet you posted still hangs for you even when you run python 3.9 and upgrade to v1.8.2 of wavefront_sdk_python.  I am scratching my head because the code snippet you posted finishes for me in my environment. I am running python 3.9.9 and using v1.8.2 of wavefront_sdk_python. There is definitely more investigation to be done. I may see if someone else on my team can reproduce the issue in their environment.

**Comment by keep94** [comment]
Just for fun, I tried stopping my wavefront proxy on my local machine. Then I re-ran your code snippet. With the proxy stopped, your code snippet does not finish for me.  Instead, it prints a stack trace and then periodically prints:   ```WARNING:wavefront_sdk.WavefrontSdkMetricsRegistry:Unable to send internal SDK metric.```. Is this the behavior you see?  When you run your code snippet in your environment, is your wavefront proxy running on your local machine?

**Comment by alexey-stankevich** [comment]
Yes, this is exactly the behavior. It's strange that it works for you with proxy running, given that I deliberately put `host='broken'`. Is there some kind of fallback to the `localhost`?

**Comment by keep94** [comment]
No, I thought you had put 'broken' in as a placeholder, so I replaced it with 'localhost'. I didn't realize it was literal.  I think I am starting to understand the issue. If the library can't connect to a wavefront proxy, it hangs, but the expected behavior is that it should eventually terminate (halt).

**Comment by keep94** [comment]
If I set max_queue_size=50000 and run the code snippet so that it cannot connect to the wavefront proxy by setting host='broken', it terminates immediately without throwing an error.  When I run the code snippet it looks like everything worked even though it could not connect to the wavefront proxy.  Is this the expected behavior when max_queue_size=1?

**Comment by alexey-stankevich** [comment]
Yes, with max_queue_size=50000 it should finish normally. The issues start when you try to push more spans than this limit (50k in case of default size). After this the call to the span.close() blocks until the queue is consumed, which will never happen in case of proxy not available. This also blocks the calling application.

The change I've done in the  #60 seems to be an expected behavior from the start, as there's a catch of the `queue.Full` exception, which can't be thrown in the blocking mode.

**Comment by keep94** [comment]
Hello. We have merged your pull request in and have published a new release, v2.3.4.  Closing this issue.
